examples/graph/mpnn-molecular-graphs.py

```diff
@@ -0,0 +1,723 @@
+        tf.debugging.assert_greater_equal(
+            x=self.batch_size,
+            y=num_molecules,
+            message="batch_size has to be >= tf.reduce_max(atom_partition_indices) + 1",
+        )
+
+        # Initialize states
+        memory_state = tf.zeros((num_molecules, self.atom_dim))
+        carry_state = tf.zeros((num_molecules, self.atom_dim))
+
+        # Perform a number of lstm steps (via the set-to-set procedure)
+        for i in range(self.steps):
+
+            # Expand carry state (to match atom_features dim)
+            carry_state_expanded = tf.gather(carry_state, atom_partition_indices)
+
+            # Perform a linear transformation followed by reduction
+            atom_features_reduced = tf.reduce_sum(
+                atom_features * carry_state_expanded, axis=1
+            )
+
+            # Split into parts correspoding to each molecule in the batch
+            atom_features_partitioned = tf.dynamic_partition(
+                atom_features_reduced, atom_partition_indices, self.batch_size
+            )
+
+            # Compute attention coefficients
+            attention_coef = tf.concat(
+                [tf.nn.softmax(features) for features in atom_features_partitioned],
+                axis=0,
+            )
+
+            # Apply attention to atom_features, and sum based on atom_partition_indices
+            attention_readout = tf.math.segment_sum(
+                tf.reshape(attention_coef, [-1, 1]) * atom_features,
+                atom_partition_indices,
+            )
+
+            # Concatenate (previous) carry_state and attention readout
+            carry_state_evolved = tf.concat([carry_state, attention_readout], axis=1)
+
+            # Perform a LSTM step (with only a memory state and carry state)
+            memory_state, carry_state = self.lstm_cell(
+                [memory_state, carry_state_evolved]
+            )
+
+        return carry_state_evolved
+
+
+"""
+### Message Passing Neural Network (MPNN)
+
+It is now time to complete the MPNN model. In addition to the message passing
+and readout, a two-layered classification network will be implemented to make
+predictions of BBBP.
+"""
+
+
+def MPNNModel(
+    atom_dim,
+    bond_dim,
+    batch_size=32,
+    message_units=64,
+    message_steps=4,
+    lstm_steps=8,
+    dense_units=256,
+):
+
+    atom_features = keras.layers.Input(
+        shape=(atom_dim), dtype="float32", name="atom_features"
+    )
+
+    bond_features = keras.layers.Input(
+        shape=(bond_dim), dtype="float32", name="bond_features"
+    )
+
+    pair_indices = keras.layers.Input(shape=(2), dtype="int32", name="pair_indices")
+
+    atom_partition_indices = keras.layers.Input(
+        shape=(), dtype="int32", name="atom_partition_indices"
+    )
+
+    x = MessagePassing(message_units, message_steps)(
+        [atom_features, bond_features, pair_indices]
+    )
+
+    x = SetGather(batch_size, lstm_steps)([x, atom_partition_indices])
```

Comment:
Try replacing this with a Transformer like https://keras.io/examples/nlp/neural_machine_translation_with_transformer/#building-the-model but without positional embedding

Reply:
Thanks for helping out. I will study your tutorial and see what I can do. I think a transformer would be a great replacement. Notice however, that I tried to follow the original paper: Do you still think I should try to replace the set-to-set (*SetGather*) layer with a transformer?

Reply:
I'm not entirely sure what the *inputs* should be to the decoder. I found [this paper](https://arxiv.org/pdf/1810.00825.pdf) which I think has an answer. They seem to use a learnable vector *S* (referred to as  *learnable set of k seed vectors* in the paper) as the *inputs*. What do you think? Do you have any other suggestions? What I want the transformer to output is a 2-D tensor (batch_size, embed_dim) which can be fed to the classification layers.

Another possible issue is that each example in the batch (which is merged into a big global graph before being fed to the MPNN) is of different size (i.e., the number of nodes differ between examples (molecules)). I'm thinking, should I obtain each example via `tf.dynamic_partition(global_graph, atom_partition_indices, ...)` and then pad each example to obtain a "rectangular" (3-D) tensor?

Reply:
> Perhaps then the decoder could be excluded (so only encoder -> pooling -> dense), similar to this example (but without position encoding/embedding).

Yes, this is a fine solution. You can try it and see if you get improved results.

> Another possible issue is that each example in the batch (which is merged into a big global graph before being fed to the MPNN) is of different size

Didn't this issue already exist in the previous implementation? Your number of steps was hardcoded. Unless I missed something.

Using padding seems like a good idea. Remember to use masking so that the padded locations don't interfere with training.